### PR TITLE
feat: refactor

### PR DESCRIPTION
**Pull request by Gorter-dev**

### CODE REVIEW

File: packages/flutter_chat/lib/src/screens/chat_profile_screen.dart [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                        ),
+                      ],
+                    ),
+                  ],
+                ),
+              ),
+            ],
+          ],
+        ),
+        if (user != null && user!.id != currentUser) ...[
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/chat_profile_screen.dart:247]: user?.id != currentUser

File: packages/flutter_chat/lib/src/screens/creation/new_chat_screen.dart [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  final FocusNode _textFieldFocusNode = FocusNode();
+  bool _isSearching = false;
+  String query = "";
+
+  @override
+  Widget build(BuildContext context) {
+    var theme = Theme.of(context);
+
+    return widget.chatOptions.builders.newChatScreenScaffoldBuilder?.call(
+          _AppBar(
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/creation/new_chat_screen.dart:38]: extract these widgets to variables

File: packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+
+  final ChatOptions options;
+  final List<UserModel> users;
+  final Function(List<UserModel> users, String chatName, String description,
+      Uint8List? image) onComplete;
+
+  @override
+  Widget build(BuildContext context) {
+    var theme = Theme.of(context);
+
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart:25]: just use an if statement

File: packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+  State<_Body> createState() => _BodyState();
+}
+
+class _BodyState extends State<_Body> {
+  final TextEditingController _chatNameController = TextEditingController();
+  final TextEditingController _bioController = TextEditingController();
+  Uint8List? image;
+
+  var formKey = GlobalKey<FormState>();
+  var isPressed = false;
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart:96]: types?

File: packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+
+  var users = <UserModel>[];
+
+  @override
+  void initState() {
+    users = widget.users;
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart:102]: other way around

**Review by freekvandeven — CHANGES_REQUESTED** [review]
i will do another review at some later point

File: packages/flutter_chat/lib/src/screens/chat_detail_screen.dart [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+  Widget build(BuildContext context) {
+    var theme = Theme.of(context);
+
+    return AppBar(
+      iconTheme: theme.appBarTheme.iconTheme ??
+          const IconThemeData(color: Colors.white),
+      centerTitle: true,
+      leading: GestureDetector(
+        onTap: () {
+          Navigator.popUntil(context, (route) => route.isFirst);
```
Comment (FlutterJoey) [inline packages/flutter_chat/lib/src/screens/chat_detail_screen.dart:123]: This is only allowed if there is a root navigator specific for this user story. Otherwise you might pop the routes that are outside the scope of this user story.

File: packages/flutter_chat/lib/src/screens/chat_detail_screen.dart [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+
+  @override
+  State<_Body> createState() => _BodyState();
+}
+
+class _BodyState extends State<_Body> {
+  ScrollController controller = ScrollController();
+  bool showIndicator = false;
+  late int pageSize;
+  var page = 0;
```
Comment (FlutterJoey) [inline packages/flutter_chat/lib/src/screens/chat_detail_screen.dart:173]: Make class properties private if they do not change.

File: packages/flutter_chat/lib/src/screens/chat_detail_screen.dart [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+  final Function(Uint8List image) onUploadImage;
+  final Function(String message) onMessageSubmit;
+  final Function(ChatModel chat) onReadChat;
+
+  @override
+  State<_Body> createState() => _BodyState();
+}
+
+class _BodyState extends State<_Body> {
+  ScrollController controller = ScrollController();
```
Comment (FlutterJoey) [inline packages/flutter_chat/lib/src/screens/chat_detail_screen.dart:170]: this can be made final

Comment (FlutterJoey) [reply packages/flutter_chat/lib/src/screens/chat_detail_screen.dart:170]: This also needs to be disposed

File: packages/flutter_chat/lib/src/screens/chat_detail_screen.dart [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+                    }
+
+                    var messages = snapshot.data?.reversed.toList() ?? [];
+
+                    WidgetsBinding.instance.addPostFrameCallback((_) async {
+                      await widget.onReadChat(widget.chat);
+                    });
+
+                    return Listener(
+                      onPointerMove: (event) {
```
Comment (FlutterJoey) [inline packages/flutter_chat/lib/src/screens/chat_detail_screen.dart:211]: extract functions to adhere to create a readable layout

**Review by FlutterJoey — CHANGES_REQUESTED** [review]
Maybe we should have the base screen builders be changed to a single builder, that provides some context as to what screen you are viewing.

The implementer can use switch expressions with pattern matching to walk through all possible options of a sealed base class to handle each screen, or not give a damn and just have the same base screen regardless of the context.

Not a finished review. Only seen up to the chat detail screen

File: packages/chat_repository_interface/.gitignore [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+# IntelliJ related
+*.iml
+*.ipr
+*.iws
+.idea/
+
+# The .vscode folder contains launch configuration and tasks you configure in
+# VS Code which you may wish to be included in version control, so this line
+# is commented out by default.
+#.vscode/
```
Comment (freekvandeven) [inline packages/chat_repository_interface/.gitignore:22]: uncomment this line

File: packages/chat_repository_interface/README.md
Comment (freekvandeven) [inline packages/chat_repository_interface/README.md:1]: Readme is still the default

File: packages/chat_repository_interface/analysis_options.yaml
Comment (freekvandeven) [inline packages/chat_repository_interface/analysis_options.yaml:1]: use the iconica linter

File: packages/chat_repository_interface/lib/src/local/local_chat_repository.dart [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+  List<ChatModel> _chats = [];
+
+  @override
+  String createChat(
+      {required List<UserModel> users,
+      String? chatName,
+      String? description,
+      String? imageUrl,
+      List<MessageModel>? messages}) {
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/local/local_chat_repository.dart:52]: trailing comma

File: packages/chat_repository_interface/lib/src/local/local_chat_repository.dart [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+    return messageController.stream;
+  }
+
+  @override
+  bool sendMessage(
+      {required String chatId,
+      required String senderId,
+      String? text,
+      String? imageUrl}) {
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/local/local_chat_repository.dart:173]: trailing comma

File: packages/chat_repository_interface/lib/src/models/chat_model.dart [hunk 13]
```diff
@@ -0,0 +1,65 @@
+import 'package:chat_repository_interface/src/models/message_model.dart';
+import 'package:chat_repository_interface/src/models/user_model.dart';
+
+class ChatModel {
+  ChatModel({
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/chat_model.dart:5]: const

File: packages/chat_repository_interface/lib/src/models/chat_model.dart [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      imageUrl: imageUrl ?? this.imageUrl,
+      canBeDeleted: canBeDeleted ?? this.canBeDeleted,
+      lastUsed: lastUsed ?? this.lastUsed,
+      lastMessage: lastMessage ?? this.lastMessage,
+      unreadMessageCount: unreadMessageCount ?? this.unreadMessageCount,
+    );
+  }
+}
+
+extension IsGroupChat on ChatModel {
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/chat_model.dart:57]: In Wellqom a Groupchat can be with 1 or 2 persons. Depending on the difference in functionality between personal and groupchats, we would still need the option to keep something a groupchat even when everyone leaves the chat and there are just 2 people

File: packages/chat_repository_interface/lib/src/models/message_model.dart [hunk 15]
```diff
@@ -0,0 +1,37 @@
+class MessageModel {
+  MessageModel({
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/message_model.dart:2]: const

File: packages/chat_repository_interface/lib/src/models/message_model.dart [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+      text: text ?? this.text,
+      imageUrl: imageUrl ?? this.imageUrl,
+      timestamp: timestamp ?? this.timestamp,
+      senderId: senderId ?? this.senderId,
+    );
+  }
+}
+
+extension MessageType on MessageModel {
+  bool isTextMessage() => text != null;
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/message_model.dart:34]: why not make this just a getter on MessageModel?

File: packages/chat_repository_interface/lib/src/models/user_model.dart [hunk 17]
```diff
@@ -0,0 +1,31 @@
+class UserModel {
+  UserModel({
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/user_model.dart:2]: const

File: packages/chat_repository_interface/lib/src/models/user_model.dart [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    this.imageUrl,
+  });
+
+  final String id;
+  final String? firstName;
+  final String? lastName;
+  final String? imageUrl;
+}
+
+extension Fullname on UserModel {
```
Comment (freekvandeven) [inline packages/chat_repository_interface/lib/src/models/user_model.dart:15]: why not make this a getter?

File: packages/chat_repository_interface/pubspec.yaml
Comment (freekvandeven) [inline packages/chat_repository_interface/pubspec.yaml:1]: cleanup this file

File: packages/chat_repository_interface/pubspec.yaml [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+environment:
+  sdk: '>=3.4.3 <4.0.0'
+  flutter: ">=1.17.0"
+
+dependencies:
+  flutter:
+    sdk: flutter
+  
+  rxdart: any
```
Comment (freekvandeven) [inline packages/chat_repository_interface/pubspec.yaml:14]: so we adding rxdart as a dependency to a package just for the local repository interface implementation?

File: packages/firebase_chat_repository/LICENSE [hunk 20]
```diff
@@ -0,0 +1 @@
+TODO: Add your license here.
```
Comment (freekvandeven) [inline packages/firebase_chat_repository/LICENSE:1]: systemlink to the toplevel license

File: packages/firebase_chat_repository/README.md
Comment (freekvandeven) [inline packages/firebase_chat_repository/README.md:1]: remove or replace the file with a proper readme

File: packages/firebase_chat_repository/analysis_options.yaml
Comment (freekvandeven) [inline packages/firebase_chat_repository/analysis_options.yaml:1]: use the iconica component analysis options

File: packages/firebase_chat_repository/pubspec.yaml
Comment (freekvandeven) [inline packages/firebase_chat_repository/pubspec.yaml:1]: cleanup

**Review by freekvandeven — CHANGES_REQUESTED** [review]
I can't seem to even run the branch with melos bootstrap or melos run get. It complains about example and flutter_chat_view

File: packages/flutter_chat/lib/src/screens/creation/widgets/search_field.dart [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+              hintText: translations.searchPlaceholder,
+              hintStyle:
+                  theme.textTheme.bodyMedium!.copyWith(color: Colors.white),
+              focusedBorder: UnderlineInputBorder(
+                borderSide: BorderSide(
+                  color: theme.colorScheme.primary,
+                ),
+              ),
+            ),
+            style: theme.textTheme.bodySmall!.copyWith(color: Colors.white),
```
Comment (freekvandeven) [inline packages/flutter_chat/lib/src/screens/creation/widgets/search_field.dart:39]: ?

File: packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart (hunk 5 above)
Comment (Gorter-dev) [reply packages/flutter_chat/lib/src/screens/creation/new_group_chat_overview.dart:119]: no?

File: packages/chat_repository_interface/lib/src/models/chat_model.dart [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+  final String id;
+  final List<MessageModel> messages;
+  final List<UserModel> users;
+  final String? chatName;
+  final String? description;
+  final String? imageUrl;
+
+  final bool canBeDeleted;
+  final DateTime? lastUsed;
+  final MessageModel? lastMessage;
```
[the remark this answers is not in the document]
Comment (Gorter-dev) [reply packages/chat_repository_interface/lib/src/models/chat_model.dart]: since messages is now gone, no :)

File: packages/chat_repository_interface/lib/src/models/message_model.dart (hunk 16 above)
Comment (Gorter-dev) [reply packages/chat_repository_interface/lib/src/models/message_model.dart:58]: I prefer the style of extensions, but I'll make it a geter

File: packages/chat_repository_interface/lib/src/models/user_model.dart (hunk 18 above)
Comment (Gorter-dev) [reply packages/chat_repository_interface/lib/src/models/user_model.dart:31]: Prefer coding style with extensions

File: packages/firebase_chat_repository/lib/firebase_chat_repository.dart
[the remark this answers is not in the document]
Comment (Gorter-dev) [reply packages/firebase_chat_repository/lib/firebase_chat_repository.dart:1]: No, this needs to be added still. It's in draft for a reason

File: packages/flutter_chat/lib/src/screens/creation/widgets/user_list.dart [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+
+          return InkWell(
+            onTap: () async {
+              if (widget.creatingGroup) {
+                return handleGroupChatTap(user);
+              } else {
+                return handlePersonalChatTap(user);
+              }
+            },
+            child: widget.options.builders.chatRowContainerBuilder?.call(
```
[the remark this answers is not in the document]
Comment (Gorter-dev) [reply packages/flutter_chat/lib/src/screens/creation/widgets/user_list.dart:90]: Open your eyes

File: packages/firebase_chat_repository/README.md
Comment (Gorter-dev) [reply packages/firebase_chat_repository/README.md:1]: Will do, when the firebase_chat_repo gets finished

**Review by freekvandeven — CHANGES_REQUESTED** [review]

**Review by FlutterJoey — APPROVED** [review]
Approved to create a smaller workload when changing this feature over time. This is not up-to-standard yet (simply because of the largely unreviewable size). So to make smaller improvements manageable we should merge this.
